torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,479 @@
+            >>> print(hidden_states.shape)
+            torch.Size([1, 1, 2, 101, 32, 5])
+        """
+        hidden_states = []
+
+        # parse inputs
+        bsz, num_concurrent_media, n_tiles, nch, w, h = images.shape
+        bsz_and_num_concurrent_media = bsz * num_concurrent_media
+
+        # if aspect_ratio is not provided, it defaults to one tile [1,1]
+        if aspect_ratio is None:
+            aspect_ratio = torch.tensor(
+                [[1, 1]] * bsz * num_concurrent_media, device=images.device
+            )
+            if n_tiles > 1:
+                raise ValueError(
+                    f"aspect_ratio was not provided, but found n_tiles>1 for {images.shape=}. Please provide aspect_ratio."
+                )
+
+        images = images.reshape(bsz_and_num_concurrent_media * n_tiles, nch, w, h)
+        aspect_ratio = aspect_ratio.reshape(bsz_and_num_concurrent_media, 2)
+
+        # patch embeddings (tokens)
+        # A tile becomes a grid of patch_grid_size X patch_grid_size patches
+        # these patches are flatenned, and called tokens from here on.
+
+        # out: (bsz * num_concurrent_media * n_tiles, embed_dim, patch_grid_size, patch_grid_size)
+        x = self.conv(images)
+
+        # out: (bsz * num_concurrent_media, n_tiles, n_tokens, embed_dim)
+        x = x.reshape(
+            bsz_and_num_concurrent_media, n_tiles, -1, self.patches_per_tile
+        ).permute(0, 1, 3, 2)
+        bsz_and_num_concurrent_media, n_tiles, n_tokens, embed_dim = x.shape
+
+        # pre_tile_pos_embed
+        if self.pre_tile_pos_embed:
+            x = self.pre_tile_pos_embed(x, aspect_ratio)
+
+        # insert cls token
+        x = self.cls_token_embedding(x)
+        n_tokens += 1
+
+        # token_pos_embedding
+        x = self.token_pos_embedding(x, aspect_ratio)
+
+        # norm
+        x = self.ln_pre(x)
+
+        # transformer with optional hidden layer outputs
+        x = x.reshape(bsz_and_num_concurrent_media, n_tiles * n_tokens, embed_dim)
+        for layer_idx, transformer_layer in enumerate(self.transformer_layers):
+            if self.indices_return_hidden and layer_idx in self.indices_return_hidden:
+                hidden_states.append(x)
+            x = transformer_layer(x)
+        x = x.reshape(bsz_and_num_concurrent_media, n_tiles, n_tokens, embed_dim)
+
+        # norm
+        x = self.ln_post(x)
+
+        # post_tile_pos_embed
+        if self.post_tile_pos_embed:
+            x = self.post_tile_pos_embed(x, aspect_ratio)
+
+        # reshape outputs
+        x = x.reshape(bsz, num_concurrent_media, n_tiles, n_tokens, embed_dim)
+        if self.indices_return_hidden:
+            hidden_states = torch.stack(hidden_states, dim=-1)
+            hidden_states = hidden_states.reshape(
+                bsz,
+                num_concurrent_media,
+                n_tiles,
+                n_tokens,
+                embed_dim,
+                len(self.indices_return_hidden),
```

Comment:
This seems like a strange output shape. Wouldn't embeds be the last layer? Otherwise we're outputting slices of the layer embeddings.

Reply:
its awkward for me too! But we later have flatten embed_dim to be `emb_dim * n_hidden`. If the hidden states are in the last dim, then this is a simple reshape. Otherwise, i have to permute it first.

I guess permuting is more beautiful